lib/puppet/functions/lookup.rb

```diff
@@ -0,0 +1,199 @@
+#
+# The search will proceed as follows:
+# For each name given in the `name` array (or once, if it's just one name):
+#  - If an override is found, it's returned
+#  - Search and optionally merge `global` (hiera), `environment`, and `module`
+#  - Return if a value is found
+#
+# Again, for each name given in the `name` array (or once, if it's just one name):
+#  - If an extra entry is found, it's returned
+#
+# Finally, fall back to default (or a given code block).
+#
+# The `merge` strategies are:
+#  - 'hash' Perform a native Ruby Hash.merge. Arguments must be Hash.
+#  - 'unique' Append everything to an array, and do a flatten and a unique at the end. Arguments can be Array or scalar.
+#  - 'deep' Perform deep_merge on arrays and hashes using DeepMerge.deep_merge with options. Arguments must be Hash or Array
+#
+# The type specification is one of:
+#
+#   * A type in the Puppet Type System, e.g.:
+#     * `Integer`, an integral value with optional range e.g.:
+#       * `Integer[0, default]` - 0 or positive
+#       * `Integer[default, -1]` - negative,
+#       * `Integer[1,100]` - value between 1 and 100 inclusive
+#     * `String`- any string
+#     * `Float` - floating point number (same signature as for Integer for `Integer` ranges)
+#     * `Boolean` - true of false (strict)
+#     * `Array` - an array (of Data by default), or parameterized as `Array[<element_type>]`, where
+#       `<element_type>` is the expected type of elements
+#     * `Hash`,  - a hash (of default `Literal` keys and `Data` values), or parameterized as
+#       `Hash[<value_type>]`, `Hash[<key_type>, <value_type>]`, where `<key_type>`, and
+#       `<value_type>` are the types of the keys and values respectively
+#       (key is `Literal` by default).
+#     * `Data` - abstract type representing any `Literal`, `Array[Data]`, or `Hash[Literal, Data]`
+#     * `Pattern[<p1>, <p2>, ..., <pn>]` - an enumeration of valid patterns (one or more) where
+#        a pattern is a regular expression string or regular expression,
+#        e.g. `Pattern['.com$', '.net$']`, `Pattern[/[a-z]+[0-9]+/]`
+#     * `Enum[<s1>, <s2>, ..., <sn>]`, - an enumeration of exact string values (one or more)
+#        e.g. `Enum[blue, red, green]`.
+#     * `Variant[<t1>, <t2>,...<tn>]` - matches one of the listed types (at least one must be given)
+#       e.g. `Variant[Integer[8000,8999], Integer[20000, 99999]]` to accept a value in either range
+#     * `Regexp`- a regular expression (i.e. the result is a regular expression, not a string
+#        matching a regular expression).
+#   * A string containing a type description - one of the types as shown above but in string form.
+#
+# If the function is called without specifying a default value, and nothing is bound to the given name
+# an error is raised unless the option `accept_undef` is true. If a block is given it must produce an acceptable
+# value (or call `error`). If the block does not produce an acceptable value an error is
```

Comment:
replace "error" with "fail" (there is no function called error callable from puppet).